### PR TITLE
Implement Algolia Instantsearch and autocomplete feature

### LINKED ISSUE #245: Algolia instantsearch and autocomplete feature integration
This Issue aims for the completion of algolia instantsearch and autocomplete feature. 
As described in issue #242 This issue will keep track of the bellow features

- [x] Autocomplete feature 
- [x] Insantsearch
- [x] Search Result

### PR DESCRIPTION
**Pull request by sandipbhuyan**
Resolves #245

### CODE REVIEW

File: ansible/roles/web/templates/parameters.yml.j2 [hunk 1]
```diff
@@ -45,3 +45,7 @@ parameters:
     # Register at https://sentry.io
     sentry_dsn:        '{{ site_sentry_dsn }}'
     sentry_env:        '{{ site_sentry_env }}'
+
+    # Algolia Configuration
```
Comment (agathver) [inline ansible/roles/web/templates/parameters.yml.j2]: not required. We are still debating whether to move into Symfony flex or not

File: ansible/roles/web/templates/parameters.yml.j2 [hunk 2]
```diff
@@ -45,3 +45,7 @@ parameters:
...
+    env(ALGOLIA_APP_ID):  '{{ site_algolia_app_id }}'
```
Comment (agathver) [inline ansible/roles/web/templates/parameters.yml.j2]: not required. see above

File: ansible/vars/dev-vagrant.user.yml [hunk 3]
```diff
@@ -8,4 +8,4 @@
 # Require confirmation of the email address of newly registered users.
 # Requires a working mailer setup. Add SMTP credentials in
 # secrets/dev-vagrant.secrets.yml before enabling this setting.
-#site_user_email_confirmation_enabled: true
+#site_user_email_confirmation_enabled: true
```
Comment (agathver) [inline ansible/vars/dev-vagrant.user.yml]: Configure your IDE to respect eol settings from `.editorconfig`. There must be a plugin available

File: site/src/Librecores/ProjectRepoBundle/Entity/Organization.php [hunk 4]
```diff
@@ -124,6 +126,24 @@ public function __construct()
...
+     *                                        can use it to serialize objects contained within this object
+     * @param string|null         $format     The format is optionally given to be able to serialize differently
+     *                                        based on different output formats
+     * @param array               $context    Options for serializing this object
+     *
+     * @return array|string|int|float|bool
+     */
+    public function normalize(NormalizerInterface $serializer, $format = null, array $context = array())
+    {
+        return [
```
Comment (agathver) [inline site/src/Librecores/ProjectRepoBundle/Entity/Organization.php]: :+1: 
This is a nice feature

File: site/src/Librecores/ProjectRepoBundle/Entity/Organization.php [hunk 5]
```diff
@@ -124,6 +126,24 @@ public function __construct()
         $this->members  = new ArrayCollection();
     }
 
+    /**
+     * Normalizes the object into an array of scalars|arrays.
```
Comment (agathver) [inline site/src/Librecores/ProjectRepoBundle/Entity/Organization.php]: Use `@inheritdoc`

File: site/src/Librecores/ProjectRepoBundle/Entity/Project.php [hunk 6]
```diff
@@ -323,6 +325,49 @@ public function __construct()
         $this->classifications = new ArrayCollection();
     }
 
+    /**
+     * Normalizes the object into an array of scalars|arrays.
```
Comment (agathver) [inline site/src/Librecores/ProjectRepoBundle/Entity/Project.php]: https://manual.phpdoc.org/HTMLSmartyConverter/HandS/phpDocumentor/tutorial_tags.inlineinheritdoc.pkg.html

No need to repeat doc from libraries

File: site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig [hunk 7]
```diff
@@ -6,149 +6,132 @@ LibreCores Search
...
 {% endblock %}
 
 {% block pagejs %}
 {# do not use autofocus attribute to work around FF bug 712130 (FOUC) #}
+
 $(function() {
-  $("#{{ search_query_form.q.vars.id }}").focus();
+  searchFunctions();
+  options =  {
+  appId: 'ENTER_APPLICATION_ID',
```
Comment (agathver) [inline site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig]: You can transclude the `site_algolia_app_id` here using twig globals

https://stackoverflow.com/a/8360126/2328165

File: site/web/assets/js/search.js [hunk 8]
```diff
@@ -0,0 +1,171 @@
+function algoliaAutocomplete() {
+  var client = algoliasearch('ENTER_APPLICATION_ID', 'ENTER_APPLICATION_KEY')
```
Comment (agathver) [inline site/web/assets/js/search.js]: Possible approach:
Transclude Algolia settings into `window.__algoliaConfig__` in a twig template and utilize it by `window.__algoliaConfig__.applicationId`

```html
<script>
window.__algoliaConfig__ = {
  applicationId: "{{ algolia_app_id }}"
  // ... more settings
}
</script>
```

File: site/web/assets/js/search.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+  var projects = client.initIndex('projects');
+  var classifications = client.initIndex('classifications');
+  var organization = client.initIndex('organization');
+  var user = client.initIndex('user');
+
+  autocomplete('#search-form-input', {debug:true}, [
```
Comment (agathver) [inline site/web/assets/js/search.js]: Make configurable through a parameter

File: ansible/roles/web/templates/parameters.yml.j2 [hunk 10]
```diff
@@ -49,3 +49,4 @@ parameters:
     # Algolia Configuration
     env(ALGOLIA_APP_ID):  '{{ site_algolia_app_id }}'
     env(ALGOLIA_API_KEY): '{{ site_algolia_api_key }}'
+    ALGOLIA_SEARCH_KEY: '{{ site_algolia_search_key }}'
```
Comment (imphil) [inline ansible/roles/web/templates/parameters.yml.j2]: No `env()` needed on that one? Why? Can we keep the two consistent?

File: site/app/config/config.yml [hunk 11]
```diff
@@ -49,13 +49,17 @@ twig:
...
+        algolia_app_id: "%env(ALGOLIA_APP_ID)%"
+        algolia_search_key: "%ALGOLIA_SEARCH_KEY%"
 
 # Assetic Configuration
 assetic:
     debug:          "%kernel.debug%"
     use_controller: false
     bundles:
         - LibrecoresProjectRepoBundle
+        - LibrecoresSiteBundle
```
Comment (imphil) [inline site/app/config/config.yml]: How is this change related to the PR? If it's not related please open a separate PR to keep this one focused.

File: site/src/Librecores/ProjectRepoBundle/Controller/DefaultController.php [hunk 12]
```diff
@@ -94,96 +96,17 @@ public function userOrgSettingsAction(Request $request, $userOrOrganization)
...
+        $searchType = $req->get('type');
+        $searchQuery = $req->get('query');
 
-        // Search for organizations
-        if ($searchQuery->getType() === SearchQuery::TYPE_ORGS) {
-            $orgs = $this->getDoctrine()
-                ->getRepository('LibrecoresProjectRepoBundle:Organization')
-                ->findByFragment($searchQuery->getQ());
-        }
+        // If searchType is null
```
Comment (imphil) [inline site/src/Librecores/ProjectRepoBundle/Controller/DefaultController.php]: ... then what? How about "search projects by default" (or remove the comment altogether)

File: site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig [hunk 13]
```diff
@@ -1,154 +1,138 @@
...
+                <i class="fa fa-code"></i>
+                <span class="sr-only">Major language: </span>
+                {{{ _highlightResult.mostUsedLanguage.value }}}
+              </span>
+              <span class="updated">
+              <i class="fa fa-history"></i> Updated {{{ activityDetails }}}
+              </span>
+            </div>
+            <hr/>
+    {% endverbatim %}
```
Comment (imphil) [inline site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig]: The indentation of this block got messed up a bit.

File: site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig [hunk 14]
```diff
@@ -1,154 +1,138 @@
...
+  <script src="{{ asset_url }}"></script>
+  {% endjavascripts %}
 {% endblock %}
 
 {% block pagejs %}
-{# do not use autofocus attribute to work around FF bug 712130 (FOUC) #}
-$(function() {
-  $("#{{ search_query_form.q.vars.id }}").focus();
-});
+  {# do not use autofocus attribute to work around FF bug 712130 (FOUC) #}
```
Comment (imphil) [inline site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig]: is this comment still relevant?

File: site/src/Librecores/SiteBundle/Resources/views/Default/home.html.twig [hunk 15]
```diff
@@ -223,9 +224,25 @@
...
+    "assets/js/algoliasearch.min.js"
+    "assets/js/autocomplete.min.js"
+    "assets/js/search.js"
+  %}
+    <script src="{{ asset_url }}"></script>
+  {% endjavascripts %}
+{% endblock %}
+
 {% block pagejs %}
 {# do not use autofocus attribute to work around FF bug 712130 (FOUC) #}
```
Comment (imphil) [inline site/src/Librecores/SiteBundle/Resources/views/Default/home.html.twig]: Place this comment closer to the code it actually describes, please.

File: site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig [hunk 16]
```diff
@@ -1,154 +1,138 @@
...
+    {% endverbatim %}
+  </script>
 
+  {# Organization Template #}
+  <script type="text/html" id="organization-template">
+    {% verbatim %}
+          <h2><a href="/{{ name }}">{{{ _highlightResult.displayName.value }}}</a></h2>
+            <div>{{{ _highlightResult.name.value }}}</div>
+            <hr/>
+    {% endverbatim %}
```
Comment (imphil) [inline site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig]: the indentation got messed up.

File: site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig [hunk 17]
```diff
@@ -1,154 +1,138 @@
...
+  {# User Template #}
+  <script type="text/html" id="user-template">
+    {% verbatim %}
+          <h2><a href="/{{ username }}">{{{ _highlightResult.username.value }}}</a></h2>
+            <div>{{{ _highlightResult.name.value }}}</div>
+            <div><i>
+                joined on {{{ createdAt.date }}}
+              </i></div>
+            <hr/>
+    {% endverbatim %}
```
Comment (imphil) [inline site/src/Librecores/ProjectRepoBundle/Resources/views/Default/project_search.html.twig]: the indentation got messed up.

File: ansible/roles/web/templates/parameters.yml.j2 (hunk 10 above)
Comment (agathver) [reply ansible/roles/web/templates/parameters.yml.j2]: env() is used to substitute for a missing env var, where it is used in getenv. This is supposed to be used directly on the front end.

**Comment by agathver** [comment]
Barring a few indent errors, this is good to go!

**Comment by sandipbhuyan** [comment]
#248 and #249 resolved

**Comment by imphil** [comment]
squashed and merged in 550622e1363cdf6b2c6fa564023f8d8ef1799140
